lightning/src/routing/scoring.rs

```diff
@@ -460,25 +467,435 @@ impl<T: Time> Readable for ChannelFailure<T> {
 	}
 }
 
+/// [`Score`] implementation using channel success probability distributions.
+///
+/// Based on *Optimally Reliable & Cheap Payment Flows on the Lightning Network* by Rene Pickhardt
+/// and Stefan Richter [[1]]. Given the uncertainty of channel liquidity balances, probability
+/// distributions are defined based on knowledge learned from successful and unsuccessful attempts.
+/// Then the negative log of the success probability is used to determine the cost of routing a
+/// specific HTLC amount through a channel.
+///
+/// Knowledge about channel liquidity balances takes the form of upper and lower bounds on the
+/// possible liquidity. Certainty of the bounds is decreased over time using a decay function. See
+/// [`ProbabilisticScoringParameters`] for details.
+///
+/// [1]: https://arxiv.org/abs/2107.05322
+pub type ProbabilisticScorer<G> = ProbabilisticScorerUsingTime::<G, ConfiguredTime>;
+
+/// Probabilistic [`Score`] implementation.
+///
+/// # Note
+///
+/// Mixing the `no-std` feature between serialization and deserialization results in undefined
+/// behavior.
+///
+/// (C-not exported) generally all users should use the [`ProbabilisticScorer`] type alias.
+#[doc(hidden)]
+pub struct ProbabilisticScorerUsingTime<G: Deref<Target = NetworkGraph>, T: Time> {
+	params: ProbabilisticScoringParameters,
+	node_id: NodeId,
```

Comment:
I'm not thrilled with it either, but it's also need in the event handling portion. I suppose we could always skip the first hop there given the first hop won't ever be scored, IIUC. In the penalty portion, it's only used to not penalize our own channels. So a similar reasoning could be used to just skip that check. I also check against the target there for any future rebalancing cases, though. Thoughts?

Reply:
Right, I believe we don't penalize our own channels, and if we pass the `EffectiveCapacity` then we can keep that behavior by not ever penalizing based on capacity if we see `EffectiveCapacity::ExactLiquidity` (which can only ever be set for our own channels - we don't know anyone else's liquidity).

Further, I'm confused by the event handling portion, then, we do `iter::once(self.node_id).chain(path...` but then the very first line in the lop is `if source == self.node_id { continue }`, so can't we just skip the first path hop int he second zip and call it a day?

Reply:
> It's possible the user didn't pass any of their own channels to find_route, instead relying on the NetworkGraph for the first hop. So we wouldn't have EffectiveCapacity::ExactLiquidity in that case.

Right, I think in this case we *do* want to score these channels, no? If we don't know the exact capacity it could be anything, and we should probably learn whether our own channels have enough capacity or not.

> Yeah, the skip should work but note we'd still score our own channel in a rebalance. Maybe this is ok? Made the change.

Ah, I'd misunderstood why you'd mentioned rebalance here. Yea, I guess that's awkward, but it feels less awkward than making the scoring data per-node. Worst-case we learn about our own channel, don't use it for our own routing, then install the scorer onto a different node and use that learned data, which seems pretty much correct?

Reply:
> Right, I think in this case we _do_ want to score these channels, no? If we don't know the exact capacity it could be anything, and we should probably learn whether our own channels have enough capacity or not.
> 

Hmm... yeah, I suppose we should. But how would we do that in the event handling code without the sender's node id? We need it to look-up the channel liquidity and swap the offsets if needed. We could special case no-swap for the first hop case, but this order may not be accurate on the last hop in the case of a rebalance.

> Ah, I'd misunderstood why you'd mentioned rebalance here. Yea, I guess that's awkward, but it feels less awkward than making the scoring data per-node. Worst-case we learn about our own channel, don't use it for our own routing, then install the scorer onto a different node and use that learned data, which seems pretty much correct?

Yeah, should be ok.

Reply:
> Hmm... yeah, I suppose we should. But how would we do that in the event handling code without the sender's node id? 

We have it in the network graph, no? A few lines up from the only use of `source` in those loops we call `as_directed_to` which can totally return the source NodeId (or it currently does via the directed channel info, but even if we remove it it can return it).

Reply:
> We have it in the network graph, no? A few lines up from the only use of `source` in those loops we call `as_directed_to` which can totally return the source NodeId (or it currently does via the directed channel info, but even if we remove it it can return it).

Would that only be in the graph it is an announced channel?

Reply:
> Would that only be in the graph it is an announced channel?

Yes, but if its not in the graph we don't need to worry about it in the scoring of it cause the router will always get it via hints if we want to route over it.

Reply:
Gotcha. Just to be clear though, we'll end up scoring our own channels. But we'd depend on `find_route` not to call `channel_penalty_msat` on them.

Reply:
Yep, I think that's totally fine. IIUC we'd only score our own *public* channels which is doubly fine, and I think ideal. Scoring our own private channels is meh, but not a big deal if we do.